### ISSUE TITLE
Dynamic dropdown (`databound-select1.xbl`) to evaluate Choice formula (`xf:itemset ref=""`) in outer scope

### ISSUE THREAD

**Issue by avernet** [orbeon/orbeon-forms#6443]
Right now, with `instance()` we have access to instance in the local model, and, conversely, not to instances that are in the outer scope.

- The XBL uses `<xbl:content includes=":root > xf|itemset, :root > xf|item, :root > xf|choices"/>`, and I would expect the content that is included this way to always evaluate in the outer scope.
- Additionally, the parent of the `xbl:content` has an `xxbl:scope="outer"`.
- Using XSLT to include the `xf:itemset` and explicitly adding `xxbl:scope="outer"` doesn't seem to help.

[+1 from customer](https://3.basecamp.com/3600924/buckets/3237397/messages/7705555059)

- [ ] Discuss with @ebruchez if the technique I use to check if an expression evaluates in the inner or outer context is valid. If so, if `xbl:content` should always evaluate in the outer scope, why using XSLT still doesn't help.

**Comment by ebruchez** [comment]
Next step: create repro case with XForms:

- XBL component
- selection control inside
- itemset in outer scope
    - but uses `context` to inner scope value
    - `instance()` should access outer scope instance, not inner scope

**Comment by ebruchez** [comment]
Verified that in this case, while the XPath context is pointing to the inner scope, the current `model` as passed to the `XFormsFunctionContext` also points to the inner scope's model, and, therefore, instance.

As discussed with @avernet, in our implementation, in general, the context `model` is statically determined: in particular, use of the `model` attribute on nested elements causes the context model to change. However, here, when just changing *XBL scope*, this does not happen.

We argue that this should happen.

**Comment by ebruchez** [comment]
The current issue seems related to the fact that we have two levels of nesting of XBL shadow trees:
 
- `fr|databound-select1`
- `xf|select1[appearance ~= dropdown]`

If I make the nested dropdown a native `xf:select1`, the scope change works.

**Comment by ebruchez** [comment]
We handle nested outer scopes wrong. Depending on the hierarchy of `xxbl:scope="..."`, we need to do things differently. But currently, in `XBLBindingBuilder`, we only handle one level of `inner`/`outer`, hence the problem.

**Comment by ebruchez** [comment]
Sample code:

```xml
<xh:html
    xmlns:xh="http://www.w3.org/1999/xhtml"
    xmlns:xf="http://www.w3.org/2002/xforms"
    xmlns:xs="http://www.w3.org/2001/XMLSchema"
    xmlns:xxf="http://orbeon.org/oxf/xml/xforms"
    xmlns:fr="http://orbeon.org/oxf/xml/form-runner"
    xmlns:xbl="http://www.w3.org/ns/xbl"
    xmlns:xxbl="http://orbeon.org/oxf/xml/xbl"
>
    <xh:head>
        <xf:model id="my-outer-model">
            <xf:instance id="my-outer-instance">
                <_>
                    <value/>
                </_>
            </xf:instance>
            <xf:instance id="my-outer-itemset">
                <_>
                    <item label="Item 1 outer" value="1"/>
                    <item label="Item 2 outer" value="2"/>
                    <item label="Item 3 outer" value="3"/>
                </_>
            </xf:instance>
        </xf:model>
        <xbl:xbl>

            <xbl:binding
                id="fr-issue6443"
                element="fr|issue6443"
                xxbl:mode="lhha binding value selection allow-minimal-label-hint"
                xxbl:label-for="fr-issue6443"
            >
                <xbl:implementation>
                    <xf:model id="issue6443-model">
                        <xf:instance><_/></xf:instance>
                        <xf:instance id="issue6443-model-itemset">
                            <_>
                                <item label="Item 1 inner" value="1"/>
                                <item label="Item 2 inner" value="2"/>
                                <item label="Item 3 inner" value="3"/>
                            </_>
                        </xf:instance>
                    </xf:model>
                </xbl:implementation>
                <xbl:template>
                    <xf:var name="binding" value="xxf:binding('fr-issue6443')"/>
                    <xf:var name="itemset" value="instance('issue6443-model-itemset')"/>
                    <xf:var name="itemset" xxbl:scope="outer">
                        <xxf:value value="$itemset" xxbl:scope="inner"/>
                    </xf:var>

                    <xf:select1
                        ref="$binding"
                        appearance="dropdown"
                        id="issue6443-select1">
                        <xf:choices context="$itemset" xxbl:scope="outer" id="issue6443-choices">
                            <xbl:content includes=":root > xf|itemset, :root > xf|item, :root > xf|choices"/>
                        </xf:choices>
                    </xf:select1>
                </xbl:template>
            </xbl:binding>
        </xbl:xbl>
    </xh:head>
    <xh:body>
        <fr:issue6443 ref="value">
<!--            <xf:itemset ref="item">&lt;!&ndash; BUG: points to the inner itemset, not the outer itemset &ndash;&gt;-->
<!--            <xf:itemset ref="item" model="my-outer-model">&lt;!&ndash; BUG: causes "Reference to non-existing model id: `my-outer-model`" &ndash;&gt;-->
            <xf:itemset ref="instance('my-outer-itemset')/item"><!-- BUG: points to the inner itemset, not the outer itemset -->
<!--            <xf:itemset ref="xxf:instance('my-outer-itemset')/item">&lt;!&ndash; works &ndash;&gt;-->
                <xf:label ref="@label"/>
                <xf:value ref="@value"/>
            </xf:itemset>
        </fr:issue6443>
    </xh:body>
</xh:html>
```

**Comment by ebruchez** [comment]
Mmh, I seem to get the correct results with this test:

```xml
<xh:html
    xmlns:xh="http://www.w3.org/1999/xhtml"
    xmlns:xf="http://www.w3.org/2002/xforms"
    xmlns:xs="http://www.w3.org/2001/XMLSchema"
    xmlns:xxf="http://orbeon.org/oxf/xml/xforms"
    xmlns:fr="http://orbeon.org/oxf/xml/form-runner"
    xmlns:xbl="http://www.w3.org/ns/xbl"
    xmlns:xxbl="http://orbeon.org/oxf/xml/xbl"
>
    <xh:head>
        <xf:model id="my-outer-model">
            <xf:instance id="my-outer-instance"><_/></xf:instance>
        </xf:model>
        <xbl:xbl>
            <xbl:binding id="fr-gaga" element="fr|gaga">
                <xbl:template id="my-gaga-template">
                    <xf:group id="fr-gaga-inner-group"/>
                    <xf:group id="fr-gaga-outer-group" xxbl:scope="outer">
                        <xf:group id="fr-gaga-inner-group-2"/>
                    </xf:group>
                </xbl:template>
            </xbl:binding>
            <xbl:binding id="fr-toto" element="fr|toto">
                <xbl:template id="my-toto-template">
                    <xf:group id="fr-toto-inner-group"/>
                    <fr:gaga id="fr-toto-inner-gaga"/>
                    <xf:group id="fr-toto-outer-group" xxbl:scope="outer">
                        <fr:gaga id="fr-toto-outer-gaga"/>
                        <xf:group id="fr-toto-inner-group-2"/>
                    </xf:group>
                </xbl:template>
            </xbl:binding>
        </xbl:xbl>
    </xh:head>
    <xh:body>
        <fr:toto id="my-toto"/>
    </xh:body>
</xh:html>
```

Resulting scopes:

```xml
<root scope="" prefixed-id="#document" model-prefixed-id="my-outer-model" binding="false" value="false">
    OK  <model scope="" prefixed-id="my-outer-model" default-instance-prefixed-id="my-outer-instance" analyzed-binds="false">
    OK      <instance scope="" prefixed-id="my-outer-instance" model-prefixed-id="my-outer-model" binding="false" value="false"/>
    --  </model>
    OK  <toto scope="" prefixed-id="my-toto" model-prefixed-id="my-outer-model" binding="false" value="false">
    OK      <template scope="my-toto" prefixed-id="my-toto≡my-toto-template" binding="false" value="false">
    OK          <group scope="my-toto" prefixed-id="my-toto≡fr-toto-inner-group" binding="false" value="false"/>
    OK          <gaga scope="my-toto" prefixed-id="my-toto≡fr-toto-inner-gaga" binding="false" value="false">
    OK              <template scope="my-toto≡fr-toto-inner-gaga" prefixed-id="my-toto≡fr-toto-inner-gaga≡my-gaga-template" binding="false" value="false">
    OK                  <group scope="my-toto≡fr-toto-inner-gaga" prefixed-id="my-toto≡fr-toto-inner-gaga≡fr-gaga-inner-group" binding="false" value="false"/>
    OK                  <group scope="my-toto" prefixed-id="my-toto≡fr-toto-inner-gaga≡fr-gaga-outer-group" binding="false" value="false">
    OK                      <group scope="my-toto" prefixed-id="my-toto≡fr-toto-inner-gaga≡fr-gaga-inner-group-2" binding="false" value="false"/>
    --                  </group>
    --            </template>
    --        </gaga>
    OK        <group scope="" prefixed-id="my-toto≡fr-toto-outer-group" model-prefixed-id="my-outer-model" binding="false" value="false">
    OK            <gaga scope="" prefixed-id="my-toto≡fr-toto-outer-gaga" model-prefixed-id="my-outer-model" binding="false" value="false">
    OK                <template scope="my-toto≡fr-toto-outer-gaga" prefixed-id="my-toto≡fr-toto-outer-gaga≡my-gaga-template" binding="false" value="false">
    OK                    <group scope="my-toto≡fr-toto-outer-gaga" prefixed-id="my-toto≡fr-toto-outer-gaga≡fr-gaga-inner-group" binding="false" value="false"/>
    OK                    <group scope="" prefixed-id="my-toto≡fr-toto-outer-gaga≡fr-gaga-outer-group" model-prefixed-id="my-outer-model" binding="false" value="false">
    OK                        <group scope="" prefixed-id="my-toto≡fr-toto-outer-gaga≡fr-gaga-inner-group-2" model-prefixed-id="my-outer-model" binding="false" value="false"/>
    --                    </group>
    --                </template>
    --            </gaga>
    OK            <group scope="" prefixed-id="my-toto≡fr-toto-inner-group-2" model-prefixed-id="my-outer-model" binding="false" value="false"/>
    --        </group>
    --    </template>
    --</toto>
</root>

```

**Comment by ebruchez** [comment]
- [ ] Noting that `itemElementAnalysis()` is probably wrong for the itemset dependencies.

**Comment by ebruchez** [comment]
Other example:

```xml
<xh:html
    xmlns:xh="http://www.w3.org/1999/xhtml"
    xmlns:xf="http://www.w3.org/2002/xforms"
    xmlns:xs="http://www.w3.org/2001/XMLSchema"
    xmlns:xxf="http://orbeon.org/oxf/xml/xforms"
    xmlns:fr="http://orbeon.org/oxf/xml/form-runner"
    xmlns:xbl="http://www.w3.org/ns/xbl"
    xmlns:xxbl="http://orbeon.org/oxf/xml/xbl"
>
    <xh:head>
        <xf:model id="my-outer-model">
            <xf:instance id="my-outer-instance"><_/></xf:instance>
        </xf:model>
        <xbl:xbl>
            <xbl:binding id="fr-gaga" element="fr|gaga">
                <xbl:template id="my-gaga-template">
                    <xbl:content includes=":root > xf|group"/>
                </xbl:template>
            </xbl:binding>
            <xbl:binding id="fr-toto" element="fr|toto">
                <xbl:template id="my-toto-template">
                    <fr:gaga id="fr-toto-inner-gaga">
                        <xf:group id="fr-toto-outer-group" xxbl:scope="outer">
                            <xbl:content includes=":root > xf|group"/>
                        </xf:group>
                    </fr:gaga>
                </xbl:template>
            </xbl:binding>
        </xbl:xbl>
    </xh:head>
    <xh:body>
        <fr:toto id="my-toto">
            <xf:group id="my-group-content"/>
        </fr:toto>
    </xh:body>
</xh:html>
```

**Comment by ebruchez** [comment]
Variation adding top-level `<fr:gaga>` for comparison:

```xml
<xh:html
    xmlns:xh="http://www.w3.org/1999/xhtml"
    xmlns:xf="http://www.w3.org/2002/xforms"
    xmlns:xs="http://www.w3.org/2001/XMLSchema"
    xmlns:xxf="http://orbeon.org/oxf/xml/xforms"
    xmlns:fr="http://orbeon.org/oxf/xml/form-runner"
    xmlns:xbl="http://www.w3.org/ns/xbl"
    xmlns:xxbl="http://orbeon.org/oxf/xml/xbl"
>
    <xh:head>
        <xf:model id="my-outer-model">
            <xf:instance id="my-outer-instance"><_/></xf:instance>
        </xf:model>
        <xbl:xbl>
            <xbl:binding id="fr-gaga" element="fr|gaga">
                <xbl:template id="my-gaga-template">
                    <xbl:content includes=":root > xf|group"/>
                </xbl:template>
            </xbl:binding>
            <xbl:binding id="fr-toto" element="fr|toto">
                <xbl:template id="my-toto-template">
                    <fr:gaga id="fr-toto-inner-gaga">
                        <xf:group id="fr-toto-outer-group" xxbl:scope="outer">
                            <xbl:content includes=":root > xf|group"/>
                        </xf:group>
                    </fr:gaga>
                </xbl:template>
            </xbl:binding>
        </xbl:xbl>
    </xh:head>
    <xh:body>
        <fr:toto id="my-toto">
            <xf:group id="my-toto-group-content"/>
        </fr:toto>
        <fr:gaga id="my-gaga">
            <xf:group id="my-gaga-group-content"/>
        </fr:gaga>
    </xh:body>
</xh:html>
```

Result:

```xml

<root scope="" prefixed-id="#document" model-prefixed-id="my-outer-model">
    OK  <model scope="" prefixed-id="my-outer-model" default-instance-prefixed-id="my-outer-instance">
    OK      <instance scope="" prefixed-id="my-outer-instance" model-prefixed-id="my-outer-model"/>
    --  </model>
    OK  <toto scope="" prefixed-id="my-toto" model-prefixed-id="my-outer-model">
    OK      <template scope="my-toto" prefixed-id="my-toto≡my-toto-template">
    OK          <gaga scope="my-toto" prefixed-id="my-toto≡fr-toto-inner-gaga">
    OK              <template scope="my-toto≡fr-toto-inner-gaga" prefixed-id="my-toto≡fr-toto-inner-gaga≡my-gaga-template">
    BAD                 <group scope="my-toto" prefixed-id="my-toto≡fr-toto-inner-gaga≡fr-toto-outer-group">
    BAD                     <group scope="my-toto" prefixed-id="my-toto≡fr-toto-inner-gaga≡my-toto-group-content"/>
    --                  </group>
    --              </template>
    --          </gaga>
    --      </template>
    --  </toto>
    OK  <gaga scope="" prefixed-id="my-gaga" model-prefixed-id="my-outer-model">
    OK      <template scope="my-gaga" prefixed-id="my-gaga≡my-gaga-template">
    OK          <group scope="" prefixed-id="my-gaga≡my-gaga-group-content" model-prefixed-id="my-outer-model"/>
    --      </template>
    --  </gaga>
</root>
```

**Comment by ebruchez** [comment]
Simpler example:

```xml
<xh:html
    xmlns:xh="http://www.w3.org/1999/xhtml"
    xmlns:xf="http://www.w3.org/2002/xforms"
    xmlns:fr="http://orbeon.org/oxf/xml/form-runner"
    xmlns:xbl="http://www.w3.org/ns/xbl"
    xmlns:xxbl="http://orbeon.org/oxf/xml/xbl"
>
    <xh:head>
        <xf:model id="my-outer-model">
            <xf:instance id="my-outer-instance"><_/></xf:instance>
        </xf:model>
        <xbl:xbl>
            <xbl:binding id="fr-gaga" element="fr|gaga">
                <xbl:template id="my-gaga-template">
                    <xbl:content includes=":root > xf|group"/>
                </xbl:template>
            </xbl:binding>
            <xbl:binding id="fr-toto" element="fr|toto">
                <xbl:template id="my-toto-template">
                    <fr:gaga id="fr-toto-inner-gaga">
                        <xf:group id="fr-toto-outer-group" xxbl:scope="outer"/>
                    </fr:gaga>
                </xbl:template>
            </xbl:binding>
        </xbl:xbl>
    </xh:head>
    <xh:body>
        <fr:toto id="my-toto"/>
    </xh:body>
</xh:html>
```

Result:

```xml
<root scope="" prefixed-id="#document" model-prefixed-id="my-outer-model">
    OK  <model scope="" prefixed-id="my-outer-model" default-instance-prefixed-id="my-outer-instance">
    OK      <instance scope="" prefixed-id="my-outer-instance" model-prefixed-id="my-outer-model"/>
    --  </model>
    OK  <toto scope="" prefixed-id="my-toto" model-prefixed-id="my-outer-model">
    OK      <template scope="my-toto" prefixed-id="my-toto≡my-toto-template">
    OK          <gaga scope="my-toto" prefixed-id="my-toto≡fr-toto-inner-gaga">
    OK              <template scope="my-toto≡fr-toto-inner-gaga" prefixed-id="my-toto≡fr-toto-inner-gaga≡my-gaga-template">
    BAD                 <group scope="my-toto" prefixed-id="my-toto≡fr-toto-inner-gaga≡fr-toto-outer-group"/>
    --              </template>
    --          </gaga>
    --      </template>
    --  </toto>
</root>
```

**Comment by ebruchez** [comment]
One clearly failing scenario above is when there is a copy with `<xbl:content>` of an element marked `xxbl:scope="outer"`.

**Comment by avernet** [comment]
Implemented workaround, and created #6456 for the underlying issue.
